File: tests/python/test_logging.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: LicenseRef-NvidiaProprietary
#
# Public ovstage logging test: install a process-global log callback, flush, and
# clear it (works on any build), and prove a channel-prefix filter with a NONE
# default threshold suppresses delivery even while the stage does log-producing
# work. CPU-only. Tested source for the logging skill.
#
# Note: there is no public API to emit a log message on demand, so this suite does
# not assert that messages ARE delivered (some builds or workloads may legitimately
# emit none); it asserts the callback lifecycle and suppression.

import math

from ovstage import LogSeverity, PopulationDomain, flush_log, population, set_log_callback

# The counting/collecting callbacks never block, so a bounded flush cannot hang.
FLUSH_TIMEOUT_NS = 5_000_000_000  # 5 s

CUBE_USDA = """#usda 1.0
(
    defaultPrim = "World"
)
def Xform "World"
{
    def Cube "Cube"
    {
        double size = 1.0
    }
}
"""


def _populate(stage):
    """Best-effort log-producing work; a no-op without the population bridge."""
    if population.available():
        population.open_usd_from_string(
            stage, CUBE_USDA, ordinal=1, time_code=math.nan, domains=PopulationDomain.RENDERING
        )


def test_log_callback_lifecycle(stage):
    """Install, flush, and clear a log callback. Requires a live Stage to bootstrap
    the runtime; needs no population bridge. Any delivered message is well-formed."""
    seen = []

    def on_log(severity, timestamp, message):
        seen.append((severity, message))

    set_log_callback(on_log, severity=LogSeverity.VERBOSE)
    try:
        assert flush_log(FLUSH_TIMEOUT_NS) is True  # drains (nothing buffered blocks)
    finally:
        set_log_callback(None)  # clear: flushes and tears the dispatcher down

    for severity, message in seen:
        assert isinstance(severity, LogSeverity)
        assert isinstance(message, str)


def test_log_callback_channel_filter_suppresses(stage):
    """A high default threshold (NONE disables all unmatched channels) plus a
    channel filter that matches nothing delivers no message, even under log-producing
    work."""
    count = [0]

    def on_log(severity, timestamp, message):
        count[0] += 1

    # [snippet:log-callback-filter]
    # set_log_callback installs a process-global callback (requires a live Stage to
    # bootstrap the runtime). The default severity is the threshold for channels not
    # named in channel_filter; LogSeverity.NONE disables all of them. channel_filter
    # is a comma-separated "<channel>=<level>" list — here a prefix that matches no
    # real channel, so nothing is delivered. flush_log forces pending messages
    # through; passing None clears the callback.
    set_log_callback(
        on_log, severity=LogSeverity.NONE, channel_filter="this.channel.does.not.exist.42=verbose"
    )
    try:
        _populate(stage)  # log-producing work the filter must suppress
        assert flush_log(FLUSH_TIMEOUT_NS) is True
    finally:
        set_log_callback(None)
    # [/snippet:log-callback-filter]

    assert count[0] == 0

File: tests/c/test_logging.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: LicenseRef-NvidiaProprietary
//
// Public ovstage logging test: install a process-global log callback, flush, and
// clear it (works on any build), and prove a channel-prefix filter with a NONE
// default threshold suppresses delivery even while the stage does log-producing
// work. CPU-only. Tested source for the logging skill.
//
// Note: there is no public API to emit a log message on demand, so this suite does
// not assert that messages ARE delivered (a given workload may legitimately emit
// none); it asserts the callback lifecycle and suppression.

#include <ovstage/ovstage.h>

#include <gtest/gtest.h>

#include <atomic>
#include <cmath>
#include <cstring>

namespace
{

// The counting callback never blocks, so a bounded flush cannot hang.
constexpr ovstage_timeout_ns_t kFlushTimeoutNs = 5'000'000'000ull;  // 5 s

std::atomic<int> g_messageCount{ 0 };

void countMessages(ovstage_log_severity_t /*severity*/, double /*timestamp*/, ovx_string_t /*message*/,
                   void* userData)
{
    static_cast<std::atomic<int>*>(userData)->fetch_add(1, std::memory_order_relaxed);
}

ovx_string_t str(const char* s)
{
    return ovx_string_t{ s, std::strlen(s) };
}

const char kCubeUsda[] =
    "#usda 1.0\n"
    "(\n"
    "    defaultPrim = \"World\"\n"
    ")\n"
    "def Xform \"World\"\n"
    "{\n"
    "    def Cube \"Cube\"\n"
    "    {\n"
    "        double size = 1.0\n"
    "    }\n"
    "}\n";

class LoggingTest : public ::testing::Test
{
protected:
    void SetUp() override
    {
        ovstage_instance_desc_t desc{};
        desc.name = "test.ovstage.logging";
        ASSERT_EQ(ovstage_create_instance(&desc, &stage_), OVSTAGE_OK);
        g_messageCount.store(0);
    }

    void TearDown() override
    {
        // Always clear the process-global callback before tearing the stage down.
        ovstage_set_log_callback(OVSTAGE_LOG_NONE, nullptr, nullptr, nullptr);
        if (stage_)
            ovstage_destroy_instance(stage_);
    }

    // Best-effort log-producing work; a no-op if the population bridge is absent.
    void populate()
    {
        ovstage_population_enqueue_result_t pop = ovstage_population_open_usd_from_string(
            stage_, str(kCubeUsda), 1, NAN, OVSTAGE_POPULATION_DOMAIN_RENDERING);
        if (pop.status == OVSTAGE_OK)
            ovstage_population_wait_op(stage_, pop.op_index, OVSTAGE_TIMEOUT_INFINITE, nullptr);
    }

    ovstage_instance_t* stage_ = nullptr;
};

TEST_F(LoggingTest, CallbackLifecycle)
{
    // Install, flush, and clear — the callback API is usable on any build (a live
    // instance bootstraps the runtime; no population bridge needed).
    ASSERT_EQ(ovstage_set_log_callback(OVSTAGE_LOG_VERBOSE, nullptr, &countMessages, &g_messageCount), OVSTAGE_OK);
    EXPECT_EQ(ovstage_flush_log(kFlushTimeoutNs), OVSTAGE_OK);
    EXPECT_EQ(ovstage_set_log_callback(OVSTAGE_LOG_NONE, nullptr, nullptr, nullptr), OVSTAGE_OK);
}

TEST_F(LoggingTest, ChannelFilterSuppresses)
{
    // [snippet:log-callback-filter-c]
    // ovstage_set_log_callback installs a process-global callback (needs a live
    // instance to bootstrap the runtime). The severity is the threshold for
    // channels not named in the filter; OVSTAGE_LOG_NONE disables all of them. The
    // channel_filter is a comma-separated "<channel>=<level>" list — here a prefix
    // that matches no real channel, so nothing is delivered. ovstage_flush_log
    // forces pending messages through; a NULL callback clears it.
    ovx_string_t bogusFilter = str("this.channel.does.not.exist.42=verbose");
    ASSERT_EQ(ovstage_set_log_callback(OVSTAGE_LOG_NONE, &bogusFilter, &countMessages, &g_messageCount), OVSTAGE_OK);
    populate();  // log-producing work the filter must suppress
    ASSERT_EQ(ovstage_flush_log(kFlushTimeoutNs), OVSTAGE_OK);
    ovstage_set_log_callback(OVSTAGE_LOG_NONE, nullptr, nullptr, nullptr);  // clear
    // [/snippet:log-callback-filter-c]

    EXPECT_EQ(g_messageCount.load(), 0);
}

} // namespace
